pageserver/src/tenant/size.rs

```diff
@@ -306,7 +322,33 @@ pub(super) async fn gather_inputs(
                 timeline_id: timeline.timeline_id,
                 kind,
             });
-            parent += 1;
+
+            parent = segments.len() - 1;
+
+            if kind == LsnKind::LeasePoint {
+                let mut lease_parent = parent;
+                segments.push(SegmentMeta {
+                    segment: Segment {
+                        parent: Some(lease_parent),
+                        lsn: lsn.0,
+                        size: None,
+                        needed: next_gc_cutoff <= lsn,
+                    },
+                    timeline_id: timeline.timeline_id,
+                    kind: LsnKind::LeaseStart,
+                });
+                lease_parent += 1;
+                segments.push(SegmentMeta {
+                    segment: Segment {
+                        parent: Some(lease_parent),
+                        lsn: lsn.0,
+                        size: None,
+                        needed: true,
+                    },
+                    timeline_id: timeline.timeline_id,
+                    kind: LsnKind::LeaseEnd,
+                });
+            }
```

Comment:
Instead of handling leases as branches within `gather_inputs`... Why do we need these?

Reply:
I think the main difficulty with handling leases as branches is that leases do not have `TimelineId`. So initially, we thought we could fix this by assigning fake timeline id to both branches and leases. 

However, we realized that when you want to calculate logical size, you do need the timeline id.

Reply:
Added explaination into doc comments.